unixfs/src/dir/builder.rs

```diff
@@ -0,0 +1,834 @@
+        self,
+        full_path: &'a mut String,
+        block_buffer: &'a mut Vec<u8>,
+    ) -> PostOrderIterator<'a> {
+        full_path.clear();
+        block_buffer.clear();
+
+        PostOrderIterator {
+            full_path,
+            old_depth: 0,
+            block_buffer,
+            pending: vec![Visited::Descent {
+                node: self.root_builder,
+                name: None,
+                depth: 0,
+            }],
+            persisted_cids: Default::default(),
+            reused_children: Vec::new(),
+            cid: None,
+            wrap_in_directory: self.opts.wrap_in_directory,
+        }
+    }
+}
+
+#[derive(Debug)]
+enum Visited {
+    Descent {
+        node: DirBuilder,
+        name: Option<String>,
+        depth: usize,
+    },
+    Post {
+        parent_id: Option<u64>,
+        id: u64,
+        name: Option<String>,
+        depth: usize,
+        leaves: Vec<(String, Leaf)>,
+    },
+}
+
+fn update_full_path(
+    (full_path, old_depth): (&mut String, &mut usize),
+    name: Option<&str>,
+    depth: usize,
+) {
+    if depth < 2 {
+        // initially thought it might be good idea to add slash to all components; removing it made
+        // it impossible to get back down to empty string, so fixing this for depths 0 and 1.
+        full_path.clear();
+        *old_depth = 0;
+    } else {
+        while *old_depth >= depth && *old_depth > 0 {
+            // we now want to pop the last segment
+            // this would be easier with pathbuf
+            let slash_at = full_path.bytes().rposition(|ch| ch == b'/');
+            if let Some(slash_at) = slash_at {
+                full_path.truncate(slash_at);
+                *old_depth -= 1;
+            } else {
+                todo!(
+                    "no last slash_at in {:?} yet {} >= {}",
+                    full_path,
+                    old_depth,
+                    depth
+                );
+            }
+        }
+    }
+
+    debug_assert!(*old_depth <= depth);
+
+    if let Some(name) = name {
+        if !full_path.is_empty() {
+            full_path.push_str("/");
+        }
+        full_path.push_str(name);
+        *old_depth += 1;
+    }
+
+    assert_eq!(*old_depth, depth);
+}
+
+pub struct PostOrderIterator<'a> {
+    full_path: &'a mut String,
+    old_depth: usize,
+    block_buffer: &'a mut Vec<u8>,
+    // our stack of pending work
+    pending: Vec<Visited>,
+    // "communication channel" from nested entries back to their parents
+    persisted_cids: HashMap<Option<u64>, BTreeMap<String, Leaf>>,
+    reused_children: Vec<Visited>,
+    cid: Option<Cid>,
+    // from TreeOptions
+    wrap_in_directory: bool,
+}
+
+fn identity_cid(number: usize) -> Cid {
+    use multihash::{Multihash, Sha2_256};
+
+    let mh = Sha2_256::digest(&number.to_le_bytes());
+    Cid::new_v0(mh).unwrap()
+}
+
+#[derive(Debug)]
+pub enum TreeConstructionFailed {
+    // TODO: at least any quick_protobuf errors here?
```

Comment:
well turns out I cannot think of a way for quick_protobuf to fail here as the buffer is large enough.

however there can be the failure of producing a too large block.